src/io/flutter/jxbrowser/JxBrowserManager.java

```diff
@@ -0,0 +1,198 @@
+import org.jetbrains.annotations.NotNull;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+
+enum JxBrowserStatus {
+  NOT_INSTALLED,
+  INSTALLATION_IN_PROGRESS,
+  INSTALLED,
+  INSTALLATION_FAILED,
+}
+
+
+public class JxBrowserManager {
+  private static JxBrowserManager manager;
+  private static final String DOWNLOAD_PATH = FileUtils.platformPath();
+  private static final AtomicReference<JxBrowserStatus> status = new AtomicReference<>(JxBrowserStatus.NOT_INSTALLED);
+  private static final Logger LOG = Logger.getInstance(JxBrowserManager.class);
+
+  private JxBrowserManager() {}
+
+  public static JxBrowserManager get() {
+    if (manager == null) {
+      return new JxBrowserManager();
+    }
+    return manager;
+  }
+
+  public void setUp(Project project) {
+    if (!status.compareAndSet(JxBrowserStatus.NOT_INSTALLED, JxBrowserStatus.INSTALLATION_IN_PROGRESS)) {
+      // This check ensures that an IDE only downloads and installs JxBrowser once, even if multiple projects are open.
+      // If already in progress, let calling point wait until success or failure (it may make sense to call setUp but proceed).
+      // If already succeeded or failed, no need to continue.
+      return;
+    }
+
+    LOG.info(project.getName() + ": Installing JxBrowser");
+
+    final File directory = new File(DOWNLOAD_PATH);
+    if (!directory.exists()) {
+      //noinspection ResultOfMethodCallIgnored
+      directory.mkdirs();
+    }
+
+    // Check for file or loading file
+    final String[] fileNames = {JxBrowserUtils.getPlatformFileName(), JxBrowserUtils.getApiFileName(), JxBrowserUtils.getSwingFileName()};
+    boolean allDownloaded = true;
+    final List<File> files = new ArrayList<>();
+    for (String fileName : fileNames) {
+      final File file = new File(DOWNLOAD_PATH + File.separator + fileName);
+      files.add(file);
+      if (!file.exists()) {
+        allDownloaded = false;
+        break;
+      }
+    }
+
+    if (allDownloaded) {
+      LOG.info(project.getName() + ": JxBrowser platform file already exists, skipping download");
+      loadClasses(files);
+      return;
+    }
+
+    final File tempLoadingFile = new File(DOWNLOAD_PATH + File.separator + JxBrowserUtils.getLoadingFileName());
+    boolean created = false;
+    try {
+      created = tempLoadingFile.createNewFile();
+    }
+    catch (IOException e) {
+      e.printStackTrace();
+    }
+
+    if (!created) {
+      // This means another IDE already created this file and has started downloading.
+      // Wait for download to finish and then try loading again.
+      LOG.info(project.getName() + ": Waiting for JxBrowser file to download");
+      int attempts = 0;
+      while (attempts < 100) {
+        if (tempLoadingFile.exists()) {
+          try {
+            Thread.sleep(1000);
```

Comment:
you should use exponential backoff and we should display a user visible message if the download fails.

Reply:
If we don't have to manage downloads among different IDEs then we don't need this check.